### ISSUE TITLE
top-level default start_size?

### ISSUE THREAD

**Issue by molpopgen** [popsim-consortium/demes-spec#140]
Quoting from the spec:

```
start_size, end_size
In the first epoch,

at least one of the start_size or the end_size MUST be specified.

If the start_size is omitted, it shall be given the same value as the end_size.

If the end_size is omitted, it shall be given the same value as the start_size.

If the deme’s start_time is infinite, the start_size MUST have the same value as the end_size.

In subsequent epochs,

If the start_size is omitted, it shall be given the same value as the previous epoch’s end_size.

If the end_size is omitted, it shall be given the same value as the start_size.
```

What role then, if any, does a default start size have?  The  way I read the description of the first epoch is that the YAML must contain start or end size, implying that no default value can be substituted in.  Given that, the section on subsequent epochs implies that we know what to do.

**Comment by molpopgen** [comment]
I'm also reading the reference implementation at the spec website.  Unless I'm missing it, there is not a way to specify epoch defaults at the top level of a Graph?

**Comment by molpopgen** [comment]
Looking at the changes in #139, I still think that there's ambiguity here.

**Comment by molpopgen** [comment]
Digging in more, I see that this is an HDM/MDM difference.  I am obviously finding this quite confusing, so I imagine that others will as well.

**Comment by molpopgen** [comment]
I think one solution is to modify the spec text to say something like, "this section assumes that any defaults have been applied (see HDM)".  But I think that brings up another question: if epochs after than the first omit `start_size`, presumably we want the MDM language to be the protocol here: take it from the previous epoch's `end_size` and DO NOT insert the default?

**Comment by grahamgower** [comment]
The defaults text needs to be integrated into the resolution section of the HDM. There are probably other important details missing too. For now, its probably best to just read the reference implementation for how the defaults apply. We'll need to make the following points:

The defaults are applied as early as possible, i.e. if the `epoch.start_size` and/or `epoch.end_size` are set (either at the toplevel or the deme level), then the following implications apply (this may not be exhaustive, I haven't thought it all through):
* no epoch need be specified at all for a given deme, so long as all required values can be inferred. If no epochs are specified, a single epoch will be constructed during resolution. The start_size/end_size will come from the defaults, the epoch's end_time was omitted so it will get the value 0, the selfing_rate and cloning_rate will both be 0. An error is raised as usual if all the values for this epoch cannot be resolved (e.g. if no epochs were specifed and no default start_size/end_size are provided).
* If there are epochs specified, then the first thing that occurs during epoch resolution is to copy in the defaults. I.e. `epoch.start_size`/`epoch.end_size` values will be copied into a given epoch before the resolution of its start_size/end_size. The defaults effectively overwrite any other implicit behaviour that might otherwise apply (such as inheritance from the previous epoch), just as if the user had specified them for all epochs directly.

**Comment by molpopgen** [comment]
> * If there are epochs specified, then the first thing that occurs during epoch resolution is to copy in the defaults. I.e. `epoch.start_size`/`epoch.end_size` values will be copied into a given epoch before the resolution of its start_size/end_size. The defaults effectively overwrite any other implicit behaviour that might otherwise apply (such as inheritance from the previous epoch), just as if the user had specified them for all epochs directly.

Okay -- this one is subtle!  Let me think more on this.

**Comment by molpopgen** [comment]
> * If there are epochs specified, then the first thing that occurs during epoch resolution is to copy in the defaults. I.e. `epoch.start_size`/`epoch.end_size` values will be copied into a given epoch before the resolution of its start_size/end_size. The defaults effectively overwrite any other implicit behaviour that might otherwise apply (such as inheritance from the previous epoch), just as if the user had specified them for all epochs directly.

What about the following case:

* one of both sizes are specified as Epoch defaults
* some demes specify some, but not all Epochs with one or both sizes

If the "first thing" is to apply the default, then those epochs with specified sizes will get over-written, which doesn't seem quite right?

**Comment by grahamgower** [comment]
You're absolutely right. If a value is specified explicitly, the default never overwrites this. Default values are only copied in only where the corresponding fields are omitted.

**Comment by molpopgen** [comment]
> You're absolutely right. If a value is specified explicitly, the default never overwrites this. Default values are only copied in only where the corresponding fields are omitted.

Gotcha.  So the HDM resolution text should eventually say something like, "if input values are None, immediately put in defaults (if not None), then proceed to resolution according to the MDM".
